Add CI scripts

smaye81 opened the pull request:
Fixes #915

This adds a `ci` script to each project and then invokes this script as part of our CI runs.

The `ci` script incorporates all tasks to verify the validity of the project such as `generate`, `test`, `build`, `type-check`, `lint`, and `test:e2e`.

Since each project contains somewhat different scripts, the `ci` script for each project incorporates whichever of the above is defined by the project.

smaye81 commented:
@timostamm running a `checkdiff` is a bit more complicated since we have a monorepo using multiple package managers. The `yarn` projects are all failing because they are not caching the `.yarn/cache` directory and therefore keep installing Linux binaries for `buf` and `esbuild`, which then fails the `checkdiff` check. We can't make use of `setup-node`'s cache because it only allows one package manager to be specified.

We could maybe break up our current `ci.yaml` into `ci-npm.yaml`, `ci-yarn.yaml`, etc. so that we can cache these independently.

smaye81 commented:
They do suggest it if you want to use the zero-installs feature, but in reading more about it, they have a [caveat](https://yarnpkg.com/features/caching#zero-installs) about Github Actions listed there. Plus, it seems to apply to PnP only, so our Unplugged package would be out of luck (unless we committed `node_modules`, which we don't want to do).

I'm just going to add this to .gitignore for both projects since it's probably more trouble than it's worth for our current requirements.

smaye81 commented:
@timostamm should be good now finally.

timostamm reviewed (commented):
Let's use yarn version 3.6.1 consistently - in also in package.json and .yarnrc.yaml.

We should be a bit more prudent when adding gitignores for checked in files:
- react/yarn-pnp/.yarn/releases/yarn-3.6.1.cjs
- react/yarn-unplugged/.yarn/releases/yarn-3.2.1.cjs
- react/yarn-unplugged/.yarn/releases/yarn-4.0.0-rc.39.cjs

Please try using corepack instead. See here for reference: https://github.com/connectrpc/connect-query-es/pull/60

smaye81 commented:
I went with `yarn v3.63` since `yarn set version berry` updates to that version. The use of `corepack` handles removing the extraneous release files in `.yarn/releases`.

I also added a version to Remix (`pnpm`).

timostamm commented on react/yarn-unplugged/.yarnrc.yml:
```diff
@@ -1,2 +1,2 @@
-yarnPath: .yarn/releases/yarn-3.2.1.cjs
 nodeLinker: node-modules
+yarnPath: .yarn/releases/yarn-3.6.3.cjs
```
This is highly confusing when using corepack, because corepack manages package manager versions. Take a look here how it works: https://nodejs.org/api/corepack.html

This line should be removed.

smaye81 replied:
Yarn's docs recommend Corepack, but they also recommend using `yarn set version <version>` to update Yarn for a project. So, we can remove this line and the entire file in `yarn-pnp`, but anytime we run `yarn set version <version>` for either of these projects, Yarn is just going to recreate the file again and add that line back.

Similarly re: the `releases` directory, Yarn's [docs](https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored) also recommend committing this directory. There's no mention of whether this is advisable when using Corepack, but I don't see an argument against it.

timostamm commented on react/yarn-pnp/.yarnrc.yml:
```diff
@@ -1 +1 @@
-yarnPath: .yarn/releases/yarn-3.6.1.cjs
+yarnPath: .yarn/releases/yarn-3.6.3.cjs
```
The file can be removed.

timostamm commented on manage.mjs:
```diff
@@ -184,6 +195,12 @@ class PackageEnt {
+    /**
+     */
+    enableCorepack() {
+      this.run(`corepack enable`);
```
Please take a look at how corepack works. It's a one-time setup. 

Is there a specific reason we're not using the approach from https://github.com/connectrpc/connect-query-es/pull/60?

smaye81 replied:
Connect-Query is just one project using pnpm. If we run Corepack one-time for this entire repo, is it going to recursively dive down into each project and enable the package manager correctly? I assumed that since we treat each directory as a separate project, this should be run inside each one.

timostamm replied:
Yes.

From the docs:

> The Corepack proxies will find the closest [package.json](https://nodejs.org/api/packages.html#nodejs-packagejson-field-definitions) file in your current directory hierarchy to extract its ["packageManager"](https://nodejs.org/api/packages.html#packagemanager) property.
> 
> If the value corresponds to a [supported package manager](https://nodejs.org/api/corepack.html#supported-package-managers), Corepack will make sure that all calls to the relevant binaries are run against the requested version, downloading it on demand if needed, and aborting if it cannot be successfully retrieved.

timostamm reviewed (commented):
These two files are still checked in but ignored. They should be removed:

- react/yarn-pnp/.yarn/releases/yarn-3.6.3.cjs
- react/yarn-unplugged/.yarn/releases/yarn-3.6.3.cjs

timostamm commented:
@smaye81, did corepack work out?

smaye81 commented:
It seemed to. I switched our `ci.yaml` to use the same approach connect-query is using and removed it from the `manage.mjs` scripts and all the CI tests are passing. 🤷

timostamm commented on react/yarn-pnp/README.md:
```diff
@@ -8,6 +8,7 @@ Other tools used:
 
 ## Getting Started
 
+### `corepack enable`
```
Users coming to this yarn example will have yarn installed. Let's remove this instruction, and add the section about corepack from https://github.com/connectrpc/connect-query-es/pull/60/files at the bottom.

Please remove the instructions `yarn set version canary` and `yarn set version berry`.

timostamm commented on react/yarn-unplugged/README.md:
```diff
@@ -11,6 +11,7 @@ Other tools used:
 
 ## Getting Started
 
+### `corepack enable`
```
See above.

smaye81 commented:
@timostamm updated. Also added the same instructions to the `remix` project.
